ENH : Speed up TargetEncoder.transform for small batches

gautamvarmadatla opened the pull request:
**Fixes #30952**
## What does this implement/fix?

This PR adds a **small-batch fast path** for `TargetEncoder.transform` that removes the large, fixed overhead observed for tiny batches, while preserving **bit-for-bit parity** with the existing vectorized implementation for:

 * missing-like values (`None`, `np.nan`, duck-typed `pandas.NA`, `pandas.NaT`, NumPy `NaT`),
 * unseen categories (fallback to `target_mean_`),
 * multiclass outputs (columns grouped by feature with class order consistent with `classes_`).

 ### High-level idea

 * After `fit`, we **eagerly** precompute per-feature structures:

   * a dict mapping **normalized category → index in `categories_[j]`**,
   * a direct view of the learned encodings (`encodings_`) and per-feature defaults (`target_mean_` or per-class vector).
 * In `transform`, for **small batches** we do a tight loop of dictionary lookups and write the corresponding encodings directly.
 * For larger inputs, we fall back to the **unchanged vectorized path**.
 * The cutoff is a **private heuristic** (`_small_batch_threshold`) and does not change the public API.

 ### Correctness details (parity)

 To match the vectorized reference exactly, `transform` normalizes keys via `_norm_key(x)` using distinct, hashable enum sentinels:

 * `None` → `_NAKey.NONE`
 * float / NumPy NaN → `_NAKey.NAN`
 * duck-typed `pandas.NA` → `_NAKey.PD_NA`
 * duck-typed `pandas.NaT` → `_NAKey.PD_NAT`
 * NumPy `datetime64`/`timedelta64` `NaT` → `_NAKey.NAT`

 The same normalization is applied when building the maps during `fit` and when looking up during `transform`. Everything else is used as-is. Unseen categories map to per-feature defaults (`target_mean_` or the per-class mean vector). In the multiclass case, columns are written feature-major with class order aligned to `classes_`.

---

## Benchmarks

**Environment**: Windows, CPython 3.12 (64-bit), NumPy MKL, `pytest-benchmark`.

**Setup:** I benchmarked both paths ( Fast and vec ) for **both dtypes**: `int64` and `object`. Here fast = per-row lookups into a precomputed map (Basically new logic foer small batches ) and vec = fully vectorized NumPy with more setup cost that win on larger batches. ( Current sklearn implementation ) . I basically forced data to explicitly flow through these paths. 

**Data & procedure (per run):**
* **Fit:** 500,000 rows with ~100,000 categories (integer-coded), binary target.
* **Transform batches:** sizes sweep over {1, 2, 4, …, 256,000}. 
* **Dtypes:**
  * `int64`: strictly integral, no missing sentinels.
  * `object`: Python objects; when `n_small ≥ 2`, inject `None` (first row) and `np.nan` (last row).
* **Paths:**
  * **fast:** `_small_batch_threshold = 1e9` (forces fast path).
  * **vec:** `_small_batch_threshold = 0` (forces vectorized path).
* **RNG:** `seed = n_small` for reproducibility.

I ran separate benchmarks for int64 and object because they exercise very different code paths. With int64, NumPy gives a tight, contiguous buffer and the vectorized implementation is highly efficient, so the fast-path only wins for tiny batches and the crossover comes early (around 2^12 rows on my machine). With object dtype, we're dealing with python pointers and mixed NA sentinels, so normalization and lookups are inherently pricier and cache-unfriendly; the small-batch path stays competitive much longer, and the crossover shifts way out (around 2^18 rows).

<img width="1189" height="989" alt="output" src="https://github.com/user-attachments/assets/5f3465e2-8dab-4f01-85af-88d978adfc45" />


<img width="1189" height="989" alt="op3" src="https://github.com/user-attachments/assets/b6742ca0-e65c-4ce2-a3cf-78eb6a1cdb1e" />

 ### Notes ( Threshold selection Logic ) 

 *  I’m keeping _small_batch_threshold as a fixed number for now. The point where the fast path wins jumps around a lot depending on dtype (object vs int64), our CPU/cache, NumPy build, and Python version, so an “adaptive” rule might pick the wrong cutoff on someone else’s machine. A conservative constant would keep the performance much more predictable and still give the small-batch speedups. So, I’m using a nujmber ( 2^10) slightly lower than the int64 break-even (~2^12 on my machine as we can see in plots ) because integer-coded categories are super common in practice (factorize, Categorical.codes, IDs stored as ints) and that crossover we see in above plots is much earlier. At 2^10 as we can see, we see significant gains for both int64 and object dtypes. I’m not using the object-array crossover (it’s much higher) because object dtype has extra Python overhead (boxing, hashing, NA handling), and its threshold swings more with hardware/software. 
 
* So, TL;DR: fixed threshold of ~2^10 for now, but happy to discuss and refine a adaptive heuristic.

---

## Line-level profiling

*  FAST
<img width="731" height="351" alt="prof1" src="https://github.com/user-attachments/assets/ffc3212e-98ce-4651-a7ec-918391b88350" />

* VEC 
<img width="693" height="351" alt="prof2" src="https://github.com/user-attachments/assets/bc1599e2-2962-4a12-8592-6062da19cb07" />


---

## Benchmarks and Lineprofiler helper code ( Latency ) 

<details>
<summary><strong>Benchmark harness (pytest-benchmark)</strong></summary>

```python
# tests_local/test_target_encoder_bench_local.py
import os, io, time, cProfile, pstats, tracemalloc
from pathlib import Path
import numpy as np, pytest
from sklearn.preprocessing import TargetEncoder

pytestmark = pytest.mark.benchmark(group="TargetEncoder-small-batch")

def _make_data(n_fit=300_000, n_cats=200_000, n_small=1, seed=0):
    rng = np.random.default_rng(seed)
    X_fit = rng.integers(0, n_cats, size=(n_fit, 1)).astype(object)
    y_fit = rng.integers(0, 2, size=(n_fit,))
    X_small = rng.integers(0, n_cats * 2, size=(n_small, 1)).astype(object)
    if n_small >= 2:
        X_small[0, 0] = None
        X_small[-1, 0] = np.nan
    return X_fit, y_fit, X_small

@pytest.mark.parametrize("n_small", [1, 2, 16, 32, 64, 128, 255, 256, 257, 512, 1024, 2048])
def test_te_adaptive_threshold_benchmark(benchmark, n_small):
    Xf, yf, Xs = _make_data(n_small=n_small, seed=n_small)
    enc = TargetEncoder().fit(Xf, yf)
    benchmark(enc.transform, Xs)
```

</details>

<details>
<summary><strong>Line-profiler helper (top-k hottest lines in ms)</strong></summary>

```python
# notebooks or local script
import numpy as np, pandas as pd, inspect, linecache
from line_profiler import LineProfiler
from tabulate import tabulate
from sklearn.preprocessing import TargetEncoder

def make_data(n_fit=300_000, n_cats=200_000, n_small=256, seed=0):
    rng = np.random.default_rng(seed)
    X_fit = rng.integers(0, n_cats, size=(n_fit, 1)).astype(object)
    y_fit = rng.integers(0, 2, size=(n_fit,))
    X = rng.integers(0, n_cats * 2, size=(n_small, 1)).astype(object)
    if n_small >= 2:
        X[0, 0] = None
        X[-1, 0] = np.nan
    return X_fit, y_fit, X

def profile_te_transform_top_ms(enc, X, top_k=3, warmup_n=32, verbose=True):
    f = getattr(TargetEncoder, "transform")
    f = getattr(f, "__wrapped__", f)  # bypass sklearn wrapper if present
    src_file = inspect.getsourcefile(f)
    if not src_file:
        raise RuntimeError("Could not locate Python source for TargetEncoder.transform")
    f(enc, X[: min(warmup_n, len(X))])  # warmup
    lp = LineProfiler(); lp.add_function(f); lp(f)(enc, X)
    stats = lp.get_stats(); sec_per_tick = float(stats.unit)
    if verbose:
        print(f"[line_profiler] seconds per tick: {sec_per_tick:.9g}")
    rows = []
    for (filename, _start, _name), timing_list in stats.timings.items():
        if filename != src_file:
            continue
        for lineno, hits, ticks in timing_list:
            rows.append({
                "line": lineno,
                "code": linecache.getline(src_file, lineno).rstrip(),
                "time_ms": (ticks * sec_per_tick) * 1e3,
            })
    if not rows:
        raise RuntimeError("No per-line timings captured. Is line_profiler installed?")
    df = (pd.DataFrame(rows)
            .groupby(["line","code"], as_index=False)["time_ms"].sum()
            .sort_values("time_ms", ascending=False)
            .head(top_k)
            .reset_index(drop=True))
    table = tabulate(
        [[r["line"], r["code"], f'{r["time_ms"]:.3f}'] for _, r in df.iterrows()],
        headers=["line", "code", "time_ms"], tablefmt="github"
    )
    print(table)
    return df
```

</details>

---

### Memory benchmarks : 

**peakmem_fit**: RSS observed while running TargetEncoder.fit(X, y)
**peakmem_transform_large:** RSS during TargetEncoder.transform(X_large) on a big batch (vectorized path)
**peakmem_transform_small:**  RSS during transform(X_small) on a tiny batch (small-batch fast path)

<img width="1333" height="623" alt="image" src="https://github.com/user-attachments/assets/778a8e50-60f8-4752-a424-0ac855cfe041" />

Thanks for reviewing! Happy to make further changes if required!

gautamvarmadatla committed:
TargetEncoder adaptive small batch fast path with parity to vectorized transform

gautamvarmadatla committed:
TargetEncoder adaptive small batch fast path with parity to vectorized transform

gautamvarmadatla committed:
TargetEncoder adaptive small batch fast path with parity to vectorized transform

gautamvarmadatla committed:
TargetEncoder adaptive small batch fast path with parity to vectorized transform

gautamvarmadatla committed:
rename force_all_finite→ensure_all_finite to resolve CI warnings

gautamvarmadatla committed:
rename force_all_finite→ensure_all_finite to resolve CI warnings

gautamvarmadatla committed:
tests: add extra edge cases for TargetEncoder — Codecov patch failed; expand coverage to hit changed lines

gautamvarmadatla committed:
tests: add extra edge cases for TargetEncoder — Codecov patch failed; expand coverage to hit changed lines

gautamvarmadatla committed:
fixed lint errors in new test file

gautamvarmadatla committed:
fixed lint errors in new test file

gautamvarmadatla committed:
Updated test file to handle pandas import error in CI on Linux

gautamvarmadatla committed:
Updated test file to handle pandas import error in CI on Linux

gautamvarmadatla committed:
Updated test file to handle pandas import error in CI on Linux

gautamvarmadatla committed:
Updated test file to handle pandas import error in CI on Linux

gautamvarmadatla committed:
Azure workflow has no pandas module so CI pipeline is failing so replaced pandas with attribute based functions. Deleted coverage test files to test code coverage

gautamvarmadatla committed:
Azure workflow has no pandas module so CI pipeline is failing so replaced pandas with attribute based functions. Deleted coverage test files to test code coverage

gautamvarmadatla committed:
new tests for Codecov coverage

gautamvarmadatla committed:
new tests for Codecov coverage

gautamvarmadatla committed:
Speed up TargetEncoder.transform for tiny batches (adaptive fast-path) while preserving exact parity

gautamvarmadatla committed:
Speed up TargetEncoder.transform for tiny batches (adaptive fast-path) while preserving exact parity

gautamvarmadatla committed:
Target encoder small batch speedup

gautamvarmadatla committed:
Target encoder small batch speedup

ogrisel commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -311,30 +444,178 @@ def fit_transform(self, X, y):
+        for i in range(n_cats_j):
+            key = _norm_key(cats_j[i])
+            # Paranoia: if a collision happens (shouldn’t with distinct sentinels),
+            # disable fastpath for this feature by leaving its map as None.
```
Please explain how such a collision would happen in a real life setting.

ogrisel reviewed (commented):
I find this PR hard to review because of the overly defensive code. Please remove it.

Also, could you please use a log scale for the y axis of the benchmark results? Could you please plot both the speed-up and the throughput or latency for each code branch ?

gautamvarmadatla commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -311,30 +444,178 @@ def fit_transform(self, X, y):
+            norm_key = _norm_key  # local ref for speed
+
+            for j in range(n_features):
+                self._ensure_fastpath_structs_for_feature(j)
```
sure!

gautamvarmadatla committed:
TargetEncoder – remove defensive fast-path fallback and corresponding tests

gautamvarmadatla committed:
TargetEncoder – remove defensive fast-path fallback and corresponding tests

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

adrinjalali reviewed (changes requested):
This looks like AI generated code, and AI generated PR description.

Looking at the code itself, it looks pretty much nothing like the code we usually have in scikit-learn.

I'm inclined to close this PR. @gautamvarmadatla , please explain how this code was written, why it's written as it is, and how much of it is your code, and please avoid using AI to generate answers to this question.

For more context, see:

- https://github.com/scikit-learn/scikit-learn/blob/main/CODE_OF_CONDUCT.md
- https://adrin.info/the-cost-of-ai-in-open-source-maintenance.html

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
remove overly defensive code and associated tests

gautamvarmadatla committed:
remove overly defensive code and associated tests

gautamvarmadatla committed:
Merge branch 'target_transform_speedup_low_batch_size' of https://github.com/gautamvarmadatla/scikit-learn into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla commented:
> Actually, the main problem is not that it's AI generated or not, but rather if the can contributors properly assess the value and the code complexity / performance trade-offs involved in the changes they propose via the PR.
> 
> The fact that this PR's contributor blindly trusted their AI tool to output such a high complexity / verbose contribution (the `__setstate__` override in particular) makes me believe that future collaboration on this PR will not be productive, and I agree that we should close it.
> 
> scikit-learn maintainers are not interested in iteratively refining automated contributions, mostly written by AI tools indirectly via the GitHub PR review interface. If we want to leverage AI automation, we can do it ourselves and do the judgment or the AI output quality on our own before issuing the PR for review by other maintainers.
> 
> @gautamvarmadatla, assuming you are a real person, I would advise you to deepen your understanding of programming by refraining from using generative coding tools so liberally when attempting to contribute to open source projects. Limit your use of code assistant tools to explain code you don't understand to you (and only you), and always exercise your own judgment on the quality of its suggestions before imposing it onto others.

I have sent an email regading this :) Extremely sorry for the defensive code ( Once Code Cov CI started failing , i changed a lot of code which was not necessrayy tbh )  and AI-ish PR text. I have refactored most of the code, please let me know if it looks good now. Tysm!

adrinjalali commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -18,9 +19,51 @@
 )
 
 
+def _looks_like_pandas_na(x) -> bool:
```
all such methods are in `utils.validation.py`, we should keep such utils in one place.

gautamvarmadatla replied:
Moved it to utils.validation.py ( added them at end of file )

adrinjalali commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -207,13 +250,17 @@ def __init__(
         self.random_state = random_state
 
+        # --- private: threshold to choose fast path
+        self._small_batch_threshold = 1024
```
this breaks our API, we never set attributes other than the constructor args in `__init__`.

gautamvarmadatla replied:
_small_batch_threshold is now a class attribute (constant), so no new instance attrs are introduced at construction.

adrinjalali commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -311,30 +360,114 @@ def fit_transform(self, X, y):
+        n_classes = 0 if not is_multi else len(self.classes_)
+
+        self._te_is_multiclass_ = is_multi
+        self._te_index_maps_ = [None] * n_features
+        self._te_enc_vecs_ = [None] * n_features
+        self._te_enc_blocks_ = [None] * n_features
+        self._te_defaults_ = [None] * n_features
```
this is adding quite a bit to the memory footprint of a fitted target encoder when there are a lot of features. Also, the `te_` is unnecessary, this object _is_ a target encoder.

gautamvarmadatla replied:
Good point. I’ve removed the large fast-path caches and the te_ prefix entirely. The fitted estimator now keeps only a lightweight per-feature {category -> ordinal index} map to support the tiny-batch path, and reads encodings_ / target_mean_ directly. For multiclass, the class-by-feature “block” is assembled on the fly per call (not stored on self).
This substantially reduces the memory footprint vs. the previous *_enc_vecs_, *_enc_blocks_, and *_defaults_ arrays

adrinjalali commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -311,30 +360,114 @@ def fit_transform(self, X, y):
         """
+        # Use the fast path for small batch else fall back to the vectorized path.
+        check_is_fitted(self)
+        X_checked = validate_data(
```
we're duplicating validation here, since it's also done in `self._transform`

adrinjalali commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -311,30 +360,114 @@ def fit_transform(self, X, y):
         return X_out
 
+    def _build_fastpath_structs(self):
+        """Precompute all fast-path structures during fit/fit_transform."""
```
It's not really clear to me what this function is supposed to do

adrinjalali commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -311,30 +360,114 @@ def fit_transform(self, X, y):
     def transform(self, X):
         """Transform X with the target encoding.
 
-        .. note::
-            `fit(X, y).transform(X)` does not equal `fit_transform(X, y)` because a
-            :term:`cross fitting` scheme is used in `fit_transform` for encoding.
-            See the :ref:`User Guide <target_encoder>`. for details.
```
is this not true anymore?

adrinjalali commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -311,30 +360,114 @@ def fit_transform(self, X, y):
+            dtype=None,
+            ensure_all_finite="allow-nan",
+        )
+        n_samples = X_checked.shape[0]
```
if we just want the number of samples, we don't have to call validation on X, we could just call `_n_samples` on it

adrinjalali commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -311,30 +360,114 @@ def fit_transform(self, X, y):
+
+        # ---- Small-batch path (tiny inputs only) ---------------------------------
+        if use_small_batch:
+            X_arr = np.asarray(X_checked, dtype=object)
```
this branch can be factored out in a separate method for better readability here

adrinjalali commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -712,3 +714,288 @@ def test_pandas_copy_on_write():
+# ---------------------------------------------------------------------------
+
+
+class _FakeDF:
```
we don't use such things, we simply use a dataframe (pandas or polars) and we skip the test if those packages are not installed (check other similar tests please)

adrinjalali reviewed (commented):
Thanks for the improvements. We also need a benchmark on the memory footprint of this estimator when we have a lot of features.

gautamvarmadatla committed:
refactored code remove te_ , move to validation.py, remove constructor from init, remove duplciate validation update docstring and tests

gautamvarmadatla committed:
refactored code remove te_ , move to validation.py, remove constructor from init, remove duplciate validation update docstring and tests

gautamvarmadatla committed:
Merge branch 'target_transform_speedup_low_batch_size' of https://github.com/gautamvarmadatla/scikit-learn into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
transform function docstring update

gautamvarmadatla committed:
transform function docstring update

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
updated pandas tests

gautamvarmadatla committed:
updated pandas tests

gautamvarmadatla committed:
Merge branch 'target_transform_speedup_low_batch_size' of https://github.com/gautamvarmadatla/scikit-learn into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
updated pandas tests

gautamvarmadatla committed:
updated pandas tests

adrinjalali reviewed (commented):
Not sure if I see the memory benchmarks.

The code paths checking for `transform` in the other tests are now changed, need to make sure all those tests pass for both transform paths.

gautamvarmadatla committed:
Ensure validation parity between paths and update validation functions

gautamvarmadatla committed:
Ensure validation parity between paths and update validation functions

gautamvarmadatla commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -311,30 +318,123 @@ def fit_transform(self, X, y):
+            ensure_2d=True,
+            dtype=None,
+            ensure_all_finite="allow-nan",
+        )
```
hey tysm, 

1) the memory results are already in the PR description ( last image ) . But I forgot to include the ASV script, my bad. Here is the code to reproduce the same memory benchmark results :

<details>
<summary><strong>Code</strong></summary>

```python
# asv_benchmarks/bench_preprocessing_target_encoder_memory.py
import numpy as np
from sklearn.preprocessing import TargetEncoder

class TargetEncoderMemSuite:
    timeout = 300
    params = [(500, 50, 4000)]
    param_names = ["n_features", "n_categories", "n_samples"]

    def setup(self, n_features, n_categories, n_samples):
        rng = np.random.default_rng(0)
        X = rng.integers(0, n_categories, size=(n_samples, n_features)).astype(object)
        y = rng.integers(0, 2, size=n_samples)
        self.enc = TargetEncoder(smooth="auto", random_state=0).fit(X, y)
        self.X_small = X[:16]
        self.X_big = X[:4096]

    def peakmem_fit(self, n_features, n_categories, n_samples):
        TargetEncoder(smooth="auto", random_state=0).fit(
            self.X_big, np.zeros(len(self.X_big), dtype=int)
        )

    def peakmem_transform_small(self, n_features, n_categories, n_samples):
        if hasattr(self.enc, "_small_batch_threshold"):
            self.enc._small_batch_threshold = 10_000_000  # force fast path
        self.enc.transform(self.X_small)

    def peakmem_transform_large(self, n_features, n_categories, n_samples):
        if hasattr(self.enc, "_small_batch_threshold"):
            self.enc._small_batch_threshold = -1  # force vectorized path
        self.enc.transform(self.X_big)
```

</details>

This is the machine i ran it on : 

```json
{
  "version": 1,
  "context": "local_machine",
  "os": "Windows 11 (64-bit)",
  "machine": "Alienware m18 R2",
  "arch": "x86_64",
  "cpu": {
    "model": "Intel(R) Core(TM) i9-14900HX",
    "base_clock_ghz": 2.20,
    "cores": 24,
    "threads": 32
  },
  "gpu": {
    "model": "NVIDIA GeForce RTX 4090",
    "vram_gb": 16
  },
  "memory_gb": 64,
  "logical_cpus": 32
}
```

2. Verified tests for both the `transform` code paths. I basically added a temporary `conftest.py` under `sklearn/tests` locally to force the path, then ran the collected `TargetEncoder` checks twice (FAST and SLOW). TLDR; everything passed, i have attached an image below showing the same. 

```python
# sklearn/tests/conftest.py
import os, pytest
from sklearn.preprocessing import TargetEncoder

@pytest.fixture(autouse=True)
def _force_te_path(monkeypatch):
    mode = os.getenv("FORCE_TE_PATH")
    if mode == "FAST":
        monkeypatch.setattr(TargetEncoder, "_small_batch_threshold", 10_000_000, raising=False)
    elif mode == "SLOW":
        monkeypatch.setattr(TargetEncoder, "_small_batch_threshold", 0, raising=False)
```

```bat
set FORCE_TE_PATH=FAST && pytest -q -k "TargetEncoder and transform"
set FORCE_TE_PATH=SLOW && pytest -q -k "TargetEncoder and transform"
```

<img width="1896" height="914" alt="image" src="https://github.com/user-attachments/assets/ad33c2de-2dad-4fad-aec4-0e8b643a6a96" />


3. Yes absolutely! I have updated script to align them. I re-ran the tests again, they seem to gave similar results.

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
update missing value handlers

gautamvarmadatla committed:
update missing value handlers

gautamvarmadatla committed:
Merge branch 'target_transform_speedup_low_batch_size' of https://github.com/gautamvarmadatla/scikit-learn into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla commented:
hi @adrinjalali , made the changes. Added a new `is_pandas_nat` in  `sklearn.utils._missing` ( similar to `is_pandas_na` ) and 
replaced local NA checks with helpers from `sklearn.utils._missing` (`is_scalar_nan`, `is_pandas_na`).

adrinjalali reviewed (approved):
Thanks, I think this LGTM now, maybe @ogrisel wants to have a look now.

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
Apply suggestion from @adrinjalali

Co-authored-by: Adrin Jalali <adrin.jalali@gmail.com>

gautamvarmadatla committed:
Apply suggestion from @adrinjalali

Co-authored-by: Adrin Jalali <adrin.jalali@gmail.com>

gautamvarmadatla reviewed (commented):
Resolved merge conflicts and applied suggested edits

gautamvarmadatla committed:
linting fix

gautamvarmadatla committed:
CI: retrigger docs

gautamvarmadatla committed:
changelog

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

OmarManzoor commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -323,8 +330,84 @@ def fit_transform(self, X, y):
+        - Uses precomputed per-feature {category -> index} maps to do O(1) lookups.
+        """
+        # Validate here (large-batch path validates inside _transform)
+        _, n_samples, n_features = self._check_X(X, ensure_all_finite="allow-nan")
```
```suggestion
        _, n_samples, n_features = self._check_X(X, ensure_all_finite="allow-nan")
```

gautamvarmadatla replied:
done!

OmarManzoor commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -323,8 +330,84 @@ def fit_transform(self, X, y):
+            n_classes = len(self.classes_)
+            X_out = np.empty((n_samples, n_features * n_classes), dtype=np.float64)
+        else:
+            X_out = np.empty((n_samples, n_features), dtype=np.float64)
```
```suggestion
        features_out = n_features * len(self.classes_) if is_multi else n_features
        X_out = np.empty((n_samples, features_out), dtype=np.float64)
```

gautamvarmadatla replied:
done!

OmarManzoor commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -323,8 +330,84 @@ def fit_transform(self, X, y):
+                default_v = np.asarray(self.target_mean_, dtype=float)
+
+                # Local block view from encodings_ (not stored on self)
+                enc_block = np.empty((n_classes, len(self.categories_[j])), dtype=float)
```
What does this comment mean?

OmarManzoor commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -337,19 +420,34 @@ def transform(self, X):
+                  (n_samples, n_features * n_classes) Encoded representation of X.
+                  For binary and continuous targets, one column per input feature is
+                  returned. For multiclass targets, one column per (feature, class)
+                  pair is returned, with classes ordered as in ``classes_``.
```
```suggestion
        X_trans : ndarray of shape (n_samples, n_features) or
                  (n_samples, n_features * n_classes) 
            Encoded representation of X. For binary and continuous targets, one column
            per input feature is returned. For multiclass targets, one column
            per (feature, class) pair is returned, with classes ordered as in ``classes_``.
```

gautamvarmadatla replied:
done!

OmarManzoor commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -337,19 +420,34 @@ def transform(self, X):
         """
+        check_is_fitted(self)
+        # Decide path WITHOUT triggering a full validation first.
+        n_samples = _num_samples(X)
```
```suggestion
        check_is_fitted(self)
        n_samples = _num_samples(X)
```

gautamvarmadatla replied:
done!

OmarManzoor commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -337,19 +420,34 @@ def transform(self, X):
+        # ---- Large/normal batches: baseline vectorized path ----------------------
         X_ordinal, X_known_mask = self._transform(
             X, handle_unknown="ignore", ensure_all_finite="allow-nan"
         )
```
```suggestion
        X_ordinal, X_known_mask = self._transform(
            X, handle_unknown="ignore", ensure_all_finite="allow-nan"
        )
```

gautamvarmadatla replied:
done!

OmarManzoor commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -712,3 +714,216 @@ def test_pandas_copy_on_write():
+    te_fast._small_batch_threshold = 1024
+    te_vec = TargetEncoder(smooth=5.0).fit(X_fit, y)
+    # White-box: force always-vectorized path for reference.
+    te_vec._small_batch_threshold = -1
```
```suggestion
    te_fast._small_batch_threshold = 1024
    te_vec = TargetEncoder(smooth=5.0).fit(X_fit, y)
    te_vec._small_batch_threshold = -1
```

OmarManzoor commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -712,3 +714,216 @@ def test_pandas_copy_on_write():
+    `_norm_key` should:
+    - map None, float NaN, and NumPy NaT to *distinct* sentinels,
+    - return ordinary strings unchanged.
+    """
```
```suggestion
def test_norm_key_real_values_cover_nan_nat_and_strings():
    """
    `_norm_key` should:
    - map None, float NaN, and NumPy NaT to *distinct* sentinels,
    - return ordinary strings unchanged.
    """
```
Please also update this docstring. It should describe briefly in 1 or 2 sentences what the test is doing.

OmarManzoor commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -712,3 +714,216 @@ def test_pandas_copy_on_write():
+
+def test_small_batch_fastpath_matches_vectorized_with_nans_and_nats():
+    """
+    For tiny inputs, small-batch dict-lookup path should match the reference
```
```suggestion
def test_small_batch_fastpath_matches_vectorized_with_nans_and_nats():
    """
    For tiny inputs, small-batch dict-lookup path should match the reference
```

OmarManzoor commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -712,3 +714,216 @@ def test_pandas_copy_on_write():
+
+    Important detail:
+    Do NOT mix datetime64 NaT and timedelta64 NaT in the *same column*, because
+    the vectorized unknown-check cannot compare them together.
```
Why do we have this instruction in a test function?

OmarManzoor commented on sklearn/utils/validation.py:
```diff
@@ -2968,3 +2970,26 @@ def validate_data(
+    NAN = 1
+    PD_NA = 2
+    PD_NAT = 3
+    NAT = 4
```
Why do we need this `Enum`? I don't think we are using it anywhere else so might as well just assign these values directly in the function below.

gautamvarmadatla replied:
You’re right, we don’t really need a full Enum here. I’ve removed _NAKey and replaced it with a few module-level singleton objects that _normalize_na_key returns directly

OmarManzoor reviewed (commented):
Thank you for the PR @gautamvarmadatla 

Here is a first set of comments

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -337,19 +420,34 @@ def transform(self, X):
             X, handle_unknown="ignore", ensure_all_finite="allow-nan"
         )
 
-        # If 'multiclass' multiply axis=1 by num of classes else keep shape the same
```
I removed it because it was just repeating what the next few lines already show about the multiclass shape, so it felt redundant. I am happy to add the comment back if you think it is still useful.

gautamvarmadatla commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -323,8 +330,84 @@ def fit_transform(self, X, y):
+                out_col = np.empty(n_samples, dtype=float)
+                for i in range(n_samples):
+                    idx = get(norm_key(col[i]), -1)
+                    out_col[i] = enc_vec[idx] if idx >= 0 else default
```
done!

gautamvarmadatla commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -712,3 +714,216 @@ def test_pandas_copy_on_write():
+
+def _fit_pair_numpy():
+    """
+    Fit two encoders on the same tiny dataset:
```
done!

gautamvarmadatla commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -712,3 +714,216 @@ def test_pandas_copy_on_write():
+    X[1] = [float("nan"), "x"]  # float NaN branch
+    X[2] = [None, "x"]  # None branch
+    X[3] = ["u", np.datetime64("NaT")]  # datetime NaT branch
+    X[4] = ["new_unseen", "x"]  # unseen → default
```
done!

gautamvarmadatla commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -712,3 +714,216 @@ def test_pandas_copy_on_write():
+# ------------------------------
+
+
+def test_smallbatch_index_maps_exist_after_fit():
```
done!

gautamvarmadatla committed:
remove extra comments, update docstrings and remove enum

gautamvarmadatla committed:
fixed failed tests due to norm key

gautamvarmadatla committed:
sphinx error fix

gautamvarmadatla committed:
sphinx error fix

gautamvarmadatla committed:
sphinx error fix

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -197,6 +201,7 @@ class TargetEncoder(OneToOneFeatureMixin, _BaseEncoder):
         "shuffle": ["boolean"],
         "random_state": ["random_state"],
     }
+    _small_batch_threshold = 1024
```
Added a brief comment next to _small_batch_threshold

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
added comment on threshold and updated n_classes

OmarManzoor commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -323,33 +333,114 @@ def fit_transform(self, X, y):
+
+        n_classes = len(self.classes_)
+        features_out = n_features * len(self.classes_) if is_multi else n_features
+        X_out = np.empty((n_samples, features_out), dtype=np.float64)
```
It seems that `self.classes_` is only valid when `is_multi` is `True`
```suggestion
        features_out = n_features
        if is_multi:
            n_classes = len(self.classes_)
            features_out *= n_classes
        X_out = np.empty((n_samples, features_out), dtype=np.float64)
```

gautamvarmadatla replied:
yes! that was what the initial code was doing. Have updated it.

OmarManzoor commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -323,33 +333,114 @@ def fit_transform(self, X, y):
-            `fit(X, y).transform(X)` does not equal `fit_transform(X, y)` because a
-            :term:`cross fitting` scheme is used in `fit_transform` for encoding.
-            See the :ref:`User Guide <target_encoder>` for details.
-
```
Why are these changes required?

OmarManzoor reviewed (commented):
Some further feedback

gautamvarmadatla commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -323,33 +333,114 @@ def fit_transform(self, X, y):
+        ``fit(X, y).transform(X)`` does not equal :meth:`fit_transform(X, y)`
+        because a :term:`cross fitting` scheme is used in
+        :meth:`fit_transform` for encoding. See the
+        :ref:`User Guide <target_encoder>` for details.
```
As a continuation to the above issue, once i had fixed the Sphinx error. It was still throwing up an error saying -> GL07: Sections are in the wrong order. Correct order is: Parameters, Returns, Notes. So I just moved the Notes section to come after Returns

gautamvarmadatla committed:
revert docstring changes

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

OmarManzoor commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -323,33 +332,116 @@ def fit_transform(self, X, y):
+        converting raw categories to their ordinal indices. This avoids
+        duplicating the learned encodings/means in memory.
+        """
+        n_features = len(self.categories_)
```
Shouldn't we call this `n_categories` here instead and also in the docstring - `per-category`

gautamvarmadatla replied:
categories_ is per feature, so len(self.categories_) is n_features, and the per-category count is len(self.categories_[j]), so I feel the current naming/docstring is apt.

OmarManzoor commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -2,6 +2,7 @@
 import warnings
 
 import numpy as np
+import numpy.testing as npt
```
Please remove this we are importing these functions directly below
```suggestion
```

OmarManzoor commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -733,3 +735,185 @@ def test_pandas_copy_on_write():
+    try:
+        return enc.transform(X)
+    finally:
+        enc._small_batch_threshold = old_thresh
```
I don't think we need this function. You can just enforce it where you want.

gautamvarmadatla replied:
okay! replaced every use below

OmarManzoor commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -733,3 +735,185 @@ def test_pandas_copy_on_write():
+    Check that `_norm_key` maps NA-like values (None, float NaN, datetime/timedelta
+    NaT) to stable sentinel objects while leaving ordinary strings unchanged, and
+    that the sentinels for each NA family are distinct from one another.
+    """
```
This test should not be in this file

OmarManzoor commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -733,3 +735,185 @@ def test_pandas_copy_on_write():
+        X_small[0, 0] = None
+        X_small[-1, 0] = np.nan
+
+    ref = _force_slow_transform(enc, X_small)
```
Just enforce the condition directly here without using this additional function.

OmarManzoor commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -733,3 +735,185 @@ def test_pandas_copy_on_write():
+        npt.assert_array_equal(names, expected)
+    else:
+        # Fall back to x0, x1 names if feature names were not captured
+        npt.assert_array_equal(names, np.array(["x0", "x1"], dtype=object))
```
```suggestion
        assert_array_equal(names, np.array(["x0", "x1"], dtype=object))
```

OmarManzoor reviewed (commented):
Further feedback

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -733,3 +735,185 @@ def test_pandas_copy_on_write():
+
+    Z_fast = te_fast.transform(X_big)  # chooses vectorized due to size
+    Z_vec = te_vec.transform(X_big)  # always vectorized
+    npt.assert_allclose(Z_fast, Z_vec, rtol=0, atol=1e-9)
```
done!

gautamvarmadatla commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -733,3 +735,185 @@ def test_pandas_copy_on_write():
+    enc = TargetEncoder(random_state=0).fit(X_fit, y_fit)
+
+    X_small = np.array([[0], [n_cats + 123], [42], [None], [np.nan]], dtype=object)
+    ref = _force_slow_transform(enc, X_small)
```
done! replaced it

gautamvarmadatla commented on sklearn/utils/validation.py:
```diff
@@ -2923,3 +2925,26 @@ def validate_data(
+    NAN = 1
+    PD_NA = 2
+    PD_NAT = 3
+    NAT = 4
```
Yeah, I noticed that too. You are right that private Enums are not common. I used one here because we need sentinel keys that cannot collide with real category values and that stay the same after pickling. If we use ints or strings they can collide with real categories, and if we use plain object() sentinels their identity is not preserved after pickling. If you have a preferred pattern in mind that fits better here, I am happy to make the changes

gautamvarmadatla committed:
npt removed, test function removed and docstring updates

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

OmarManzoor commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -240,6 +247,7 @@ def fit(self, X, y):
             Fitted encoder.
         """
         self._fit_encodings_all(X, y)
+        self._build_index_maps()
```
I don't think we need to initialize these maps when we are not dealing with the normal case. Maybe these initializations can be added under the appropriate conditions.

gautamvarmadatla replied:
Done. yes! The main issue was dict mutation after `fit`. Previously, the small-batch path only worked if `_index_maps_` had already been created, otherwise `transform()` would fall back to the vectorized path. Now `_index_maps_` is a lazy cache that only materializes the per-feature dicts when the small-batch path is actually hit. Once built, it’s just read-only `.get()` lookups, so there’s no per-call dict mutation.

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

lesteve commented:
@gautamvarmadatla I think you are underestimating the effort involved to get this PR merged. To me this does not feel anywhere near the finish line. For example nobody looked closely at the benchmarks you did which is kind of the original point of the issue and benchmarking is hard.

Basically you picked an issue that was way too hard compared to your current understanding of the scikit-learn project. That's completely OK, sometimes it's hard to estimate the effort involved even for experienced maintainers.

What I am suggesting as I said is to try to work on easier issues, get a better understanding of the project come back to this PR in a few weeks/months. I do understand this can be disappointing but that would be the best advice I would personally give you. Other maintainers may have different opinions which are equally valuable as mine.

In the spirit of Christmas, here is a simpler issue to tackle https://github.com/scikit-learn/scikit-learn/issues/32906 :wink:. Hopefully there is enough material there to get you started. I would be willing to review your PR on this issue.

To be explicit: I won't have time to reply in the next two weeks or so, and probably other reviewers will have little availability during this holidays period.

gautamvarmadatla committed:
lazy init for maps and test code updates

gautamvarmadatla committed:
Merge branch 'scikit-learn:main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
remove unused TargetEncoder index map builder and adjust tests

GaelVaroquaux commented:
Without concluding on the technical points (I haven't taken the time to have an opinion), @gautamvarmadatla I want to stress that I appreciate how you've personally reacted to the difficulties: always polite and forward looking.
In the long term, this is really important and it enables the group to have difficult discussions and decisions, such as not merging a contribution.

gautamvarmadatla committed:
refactor core method for speedup and add benchmarks

gautamvarmadatla committed:
refactor core method for speedup and add benchmarks

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
add test for codecov

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
add memory to benchmarks

gautamvarmadatla committed:
Merge branch 'main' into target_transform_speedup_low_batch_size

gautamvarmadatla committed:
eager category maps, single dict
  loop, move _NAKey to _missing.py

adrinjalali reviewed (commented):
I think this is now minimal enough for it to be a much easier merge.

cakedev0 commented:
I took a look and I'd prefer not to merge this PR in this form, but the overall idea is good (a fast path for small batches).

I did some profiling and I found out that the main problem is that **`transform` has some O(n_categories) processing for the object path**, which indeed is a real problem for people putting models in production. It can be arbitrary slow: it's ~50ms for 100k categories, 500ms for 1M, 5s for 10M... (for `transform` on one single sample).
This is a real big problem for backend latency and optimizing it away is important, it's not like going from 0.5ms to 0.01ms for the int path (this is not a speed-up we care about honestly).

The fast-path for small batches implemented in this shortcuts both:
- `self._transform`: this is `_BaseEncoder._transform`, **it's common with one-hot & ordinal encoding**. It's where O(n_categories) happens.
- `self._transform_X_ordinal`: this is really simple and not very different from what happens in the fast path => we should rewrite this part twice.

**Conclusion:** the optimization should be done at the `_BaseEncoder._transform` level.

Also, I would prefer to find something that is not a specialized fast path, but just a nice refactor that avoids the O(n_categories) processing. That might be hard to do though, and maybe the best solution is to adapt what have been done in this PR to `_BaseEncoder._transform`.

@gautamvarmadatla : I propose to open a PR that supersedes this one, I'll credit you in the changelog and in the PR description as a co-author. Would you be ok with that?

gautamvarmadatla commented:
hi @cakedev0

Sounds good, please feel free to work on it, I currently don't have the bandwidth.

Agreed the object path is the real cost, every transform rebuilds the full category set and dict so you pay for all the categories regardless of batch size.

My only pushback is, I wouldn't write off the int path. A speedup is still a speedup at high throughput, and a proper caching fix gets both anyway. Also worth keeping in mind, caching in _BaseEncoder means OrdinalEncoder and OneHotEncoder carry the map too (betatim's memory point), so lazy on first transform beats eager in fit.

betatim commented:
Should we close this PR already or will you remember to do that when you open and link the follow up @cakedev0 ?

cakedev0 commented:
Let's close this PR and open an issue instead.

@gautamvarmadatla if you want to open an issue that would be very helpful, especially if you have an example of how you had to work around it to put a model in production.

Otherwise I'll do it.

----

I've given it a first vibe-try / investigation pass and here are my findings: 
- I checked if https://github.com/scikit-learn/scikit-learn/pull/34393 solved this problem and it doesn't, so it's worth continuing in the direction of this PR.
- The memory trade-off is indeed real when there are many categories, in which case you're more likely to benefit from the optimization. It might be possible to do a cache-like pattern that doesn't make the first transform much slower than the current speed, which would IMO solve this trade-off quite well.

lesteve commented:
Agree with https://github.com/scikit-learn/scikit-learn/pull/32368#issuecomment-4914372543, let's close this.
